### PR TITLE
Introducing "took" time (in ms) for `_msearch`

### PR DESCRIPTION
**Pull request by olcbean**
As per https://github.com/elastic/elasticsearch/issues/23131#issuecomment-280307124 it is not clear if  introducing took time for multi search is a desired behavior or if a further discussion is needed.

But as there are no comments against it so far I am submitting this PR.

### CODE REVIEW

**Comment by olcbean** [comment]
Hm...
I signed the CLA, but still the check is failing...

File: core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java [hunk 1]
```diff
@@ -63,15 +63,18 @@ public void testBatchExecute() throws Exception {
         when(actionFilters.filters()).thenReturn(new ActionFilter[0]);
         ThreadPool threadPool = new ThreadPool(settings);
         TaskManager taskManager = mock(TaskManager.class);
-        TransportService transportService = new TransportService(Settings.EMPTY, null, null, TransportService.NOOP_TRANSPORT_INTERCEPTOR,
-            boundAddress -> DiscoveryNode.createLocal(settings, boundAddress.publishAddress(), UUIDs.randomBase64UUID()), null) {
+        TransportService transportService = new TransportService(Settings.EMPTY, null, null,
+                TransportService.NOOP_TRANSPORT_INTERCEPTOR,
+                boundAddress -> DiscoveryNode.createLocal(settings, boundAddress.publishAddress(),
+                        UUIDs.randomBase64UUID()), null) {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java]: Please revert all line-length changes that are changing lines that otherwise were not impacted by the changes in this PR.

File: core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java [hunk 2]
```diff
@@ -102,14 +111,17 @@ protected void doExecute(SearchRequest request, ActionListener<SearchResponse> l
                 });
             }
         };
-        TransportMultiSearchAction action =
-                new TransportMultiSearchAction(threadPool, actionFilters, transportService, clusterService, searchAction, resolver, 10);
+
+        TransportMultiSearchAction action = new TransportMultiSearchAction(threadPool,
+                actionFilters, transportService, clusterService, searchAction, resolver, 10,
+                System::nanoTime);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java]: As far as I can tell, this is the only change in this file that was necessitated by the changes in this PR.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        clusterService = createClusterService(threadPool);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        clusterService.close();
+    }
+    
+    private enum Clock {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Please remove, a simple boolean like in the other took tests is fine.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left some nits, but they're important as reverting these changes will make the PR smaller.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java (hunk 3 above)
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: It was added to readability but I will replace it with a boolean.

**Comment by olcbean** [comment]
@jasontedor I hope I addressed all of your remarks.
Thank you for your review :)

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 4]
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
...
 
     private Item[] items;
+    
+    private long tookInMillis;
 
     MultiSearchResponse() {
     }
 
-    public MultiSearchResponse(Item[] items) {
+    public MultiSearchResponse(long tookInMillis, Item[] items) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: Let's flip the order of these parameters (in general, I think that new parameters should be added at the end absent a good reason otherwise, but putting it at the end also aligns it more closely with the `BulkResponse`; it wouldn't surprise me we are not consistent about this through the codebase, but I'm choosing the `BulkResponse` here to be consistent with).

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 5]
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
...
+        this.tookInMillis = tookInMillis;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: Let's move this parameter down after the order of the parameter is changed.

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 6]
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
     }
 
     private Item[] items;
+
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: Can you remove this unnecessary space?

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 7]
```diff
@@ -130,6 +134,20 @@ public MultiSearchResponse(Item[] items) {
     public Item[] getResponses() {
         return this.items;
     }
+    
+    /**
+     * How long the msearch took.
+     */
+    public TimeValue getTook() {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: It looks like this is only used in tests, and only to get the milliseconds, so I think that we can remove it and just use `getTookInMillis` in those places.

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 8]
```diff
@@ -138,6 +156,7 @@ public void readFrom(StreamInput in) throws IOException {
         for (int i = 0; i < items.length; i++) {
             items[i] = Item.readItem(in);
         }
+        tookInMillis = in.readVLong();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: There are backwards compatibility concerns here. Consider a mixed-version cluster that contains nodes that do not know about took. You need to guard this read in a version check on the stream. For now, let's try for on or after 5.4.0. We can adjust this later if needed.

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 9]
```diff
@@ -147,11 +166,13 @@ public void writeTo(StreamOutput out) throws IOException {
         for (Item item : items) {
             item.writeTo(out);
         }
+        out.writeVLong(tookInMillis);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: There is a backwards compatibility concern here too.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 10]
```diff
@@ -52,19 +55,23 @@ public TransportMultiSearchAction(Settings settings, ThreadPool threadPool, Tran
         this.clusterService = clusterService;
         this.searchAction = searchAction;
         this.availableProcessors = EsExecutors.numberOfProcessors(settings);
+        relativeTimeProvider = System::nanoTime;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: Can we prefix this with `this`, it will make it symmetric with the other fields, and the other constructor.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 11]
```diff
@@ -154,20 +162,32 @@ private void handleResponse(final int responseSlot, final MultiSearchResponse.It
...
+             * Builds how long it took to execute the msearch.
+             */
+            private long getExecTimeInMillis(long startTimeNanos) {
+                return TimeUnit.NANOSECONDS.toMillis(getRelativeTime() - startTimeNanos);
+            }
+
         });
     }
 
+    private long getRelativeTime() {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: In `TransportBulkAction`, the corresponding method is named `relativeTime`; let's do the same here.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 12]
```diff
@@ -110,10 +117,11 @@ static int defaultMaxConcurrentSearches(int availableProcessors, ClusterState st
      * @param responseCounter incremented on each response
      * @param listener the listener attached to the multi-search request
      */
-    private void executeSearch(
+    void executeSearch(
             final Queue<SearchRequestSlot> requests,
             final AtomicArray<MultiSearchResponse.Item> responses,
             final AtomicInteger responseCounter,
+            long startTimeInNanos,
             final ActionListener<MultiSearchResponse> listener) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: Let's change the signature of this method to accept an `ActionListener<Item[]>`, and then invoke this method by defining a new `listener` that wraps the `ActionListener<MultiSearchResponse>` and takes care of creating the `MutliSearchResponse` with the took time. This removes the need for threading `startTimeInNanos` all the way through execute search which I think simplifies things.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 13]
```diff
@@ -154,20 +162,32 @@ private void handleResponse(final int responseSlot, final MultiSearchResponse.It
...
 
             private void finish() {
-                listener.onResponse(new MultiSearchResponse(responses.toArray(new MultiSearchResponse.Item[responses.length()])));
+                listener.onResponse(new MultiSearchResponse(getExecTimeInMillis(startTimeInNanos), responses.toArray(new MultiSearchResponse.Item[responses.length()])));
             }
+            
+            /**
+             * Builds how long it took to execute the msearch.
+             */
+            private long getExecTimeInMillis(long startTimeNanos) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: In `TransportBulkAction`, the corresponding method is named `buildTookInMillis`; let's do the same here.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 14]
```diff
@@ -177,7 +197,5 @@ private void finish() {
             this.request = request;
             this.responseSlot = responseSlot;
         }
-
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: Please revert this change.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 15]
```diff
@@ -177,7 +197,5 @@ private void finish() {
...
     }
-
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: Please revert this change.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * MultiSearch took time tests
+ */
+public class MultiSearchActionTookTests extends ESTestCase {
+
+    private static ThreadPool threadPool;
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Let's make this an instance field and set it up and tear it down in test setup and teardown methods rather than using static. Yes, I think that this is a departure from how we do it in other places but I think doing it this way is better.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        clusterService = createClusterService(threadPool);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: The super teardown should be done after the specific teardown.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+        clusterService.close();
+    }
+    
+    //Unit conversion using a controller clock
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Can you add a space between the `//` and `Unit`, add the word `test` and lowercase `Unit`?

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    public void testTookWithControlledClock() throws Exception {
+        runTestTook(true);
+    }
+
+    //Using {@link System#nanoTime()}
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Let's remove the use of Javadoc syntax here since this is not a Javadoc, add a space between `//` and `Using`, add the word `test`, and lowercase `Using`.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+                            equalTo(multiSearchResponse.getTook().getMillis()));
+                } else {
+                    assertThat(multiSearchResponse.getTook().getMillis(), greaterThanOrEqualTo(
+                            TimeUnit.MILLISECONDS.convert(expected.get(), TimeUnit.NANOSECONDS)));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn(e);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: This should probably be a fail.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchIT.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.query.MatchQueryBuilder;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+/**
+ * MultiSearch IT
+ */
+public class MultiSearchIT extends ESSingleNodeTestCase {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchIT.java]: I do not think that we need an integration test for this. I appreciate that you wrote it, but I think unit test suffices.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Thanks so much @olcbean, I left a first-round review.

File: core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java (hunk 7 above)
Comment (olcbean) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: The method exists in both `SearchResponse` and `BulkResponse`.
I would prefer to keep the consistency unless there is a good reason no to.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: There is a good reason, the method is unneeded.

Comment (olcbean) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: So shall  remove also the corresponding method for `Bulk` and `Search`?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java]: Not as part of this change.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java (hunk 17 above)
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Agreed. However I was closely following the TransportBulkActionTookTests. Maybe you should change it there too?

**Comment by dakrone** [comment]
@olcbean thanks for opening this! It looks like it needs to have master merged in to fix some of the version constants, also, is it ready for another review by @jasontedor?

**Comment by olcbean** [comment]
@dakrone thanks for following up on this :) 

I am currently traveling, will rebase at the start of next week.

**Comment by olcbean** [comment]
@jasontedor I just rebased on top of the current master and set the bwc version to 7.0 ( perhaps should be changed to 6.0 ? ). I also added  a short note to the migration docs - feel free to leave it off if the change is not sufficiently notable.  

Could you please trigger the CI and let me know if any further changes are needed. 

CC-ing @dakrone as he also expressed interest in this PR.

**Comment by dakrone** [comment]
@elasticmachine test this please

**Comment by olcbean** [comment]
It has been about a month with no communication on this PR. Could you please let me know what the next steps are? @jasontedor @dakrone @rjernst

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 22]
```diff
@@ -34,16 +35,19 @@
...
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.LongSupplier;
 
 public class TransportMultiSearchAction extends HandledTransportAction<MultiSearchRequest, MultiSearchResponse> {
 
     private final int availableProcessors;
     private final ClusterService clusterService;
     private final TransportAction<SearchRequest, SearchResponse> searchAction;
+    private final LongSupplier relativeTimeProvider;
+    private SetOnce<Long> startTimeInNanos;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: I don't think `SetOnce<Long>` here is desirable, that's two extra allocations on every multi-search request.

Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: The immutability is a positive though, we can obtain this by pushing the start time all the way down via the execute method (i.e., add it as a parameter and chain it through all the way down to finish).

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a few comments.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java (hunk 22 above)
Comment (olcbean) [reply core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: @jasontedor good point! Thanks.

**Comment by olcbean** [comment]
@jasontedor thank you for picking up this PR. Can you have another look?

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 23]
```diff
@@ -123,7 +121,8 @@ void executeSearch(
             final Queue<SearchRequestSlot> requests,
             final AtomicArray<MultiSearchResponse.Item> responses,
             final AtomicInteger responseCounter,
-            final ActionListener<MultiSearchResponse> listener) {
+            final ActionListener<MultiSearchResponse> listener,
+            long startTimeInNanos) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: Please make this parameter final, and name it `relativeStartTime` (to make it clear that it's start time, but it's a relative and not absolute start time).

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 24]
```diff
@@ -72,7 +70,7 @@ public TransportMultiSearchAction(Settings settings, ThreadPool threadPool, Tran
 
     @Override
     protected void doExecute(MultiSearchRequest request, ActionListener<MultiSearchResponse> listener) {
-        startTimeInNanos = new SetOnce<>(relativeTime());
+        long startTimeInNanos = relativeTime();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: This can be final, and change it's name `relativeStartTime`.

File: core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java [hunk 25]
```diff
@@ -164,9 +165,9 @@ protected void doExecute(SearchRequest request, ActionListener<SearchResponse> l
                     availableProcessors, expected::get) {
                 @Override
                 void executeSearch(final Queue<SearchRequestSlot> requests, final AtomicArray<MultiSearchResponse.Item> responses,
-                        final AtomicInteger responseCounter, final ActionListener<MultiSearchResponse> listener) {
+                        final AtomicInteger responseCounter, final ActionListener<MultiSearchResponse> listener, long startTimeInNanos) {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java]: Same suggestion about the addition parameter.

File: core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java [hunk 26]
```diff
@@ -155,20 +162,32 @@ private void handleResponse(final int responseSlot, final MultiSearchResponse.It
...
+             * Builds how long it took to execute the msearch.
+             */
+            private long buildTookInMillis() {
+                return TimeUnit.NANOSECONDS.toMillis(relativeTime() - startTimeInNanos);
             }
         });
     }
 
+    private long relativeTime() {
+        return relativeTimeProvider.getAsLong();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java]: I don't think this method is needed.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a few more comments.

**Comment by olcbean** [comment]
@jasontedor can you have another look?

**Comment by jasontedor** [comment]
test this please

**Comment by olcbean** [comment]
@jasontedor so how does it look?

File: docs/reference/migration/migrate_7_0/java.asciidoc [hunk 27]
```diff
@@ -0,0 +1,9 @@
+[[breaking_70_java_changes]]
```
Comment (jasontedor) [inline docs/reference/migration/migrate_7_0/java.asciidoc]: We typically do not add changes here for features adds, preferring to reserve it for changes users need to know to help them migrate; I see this as a feature add. Could you please remove this?

**Review by jasontedor — CHANGES_REQUESTED** [review]
There is one more change I would like, then this can be pulled in.

**Comment by olcbean** [comment]
@jasontedor just pushed the latest changes. Take another look?

**Comment by jasontedor** [comment]
test this please

**Comment by jasontedor** [comment]
test this please

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jasontedor** [comment]
Thanks @olcbean.
